### ISSUE TITLE
Bug: Visual Basic late binder does not respect C# init properties

### ISSUE THREAD

**Issue by HaloFour** [dotnet/runtime#45039]
Applies to C# 9.0 and Visual Basic running on .NET 5.0.

The Visual Basic [late binder](https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/CompilerServices/NewLateBinding.vb) in the Microsoft.VisualBasic.Core runtime does not respect unknown `modreq`s applied to member signatures which should prevent those members from being invoked.  This is problematic in general, but more specifically it enables late binding to completely bypass the restrictions implied by `modreq(System.Runtime.CompilerServices.IsExternalInit)` used by C# 9.0 to denote `init` properties.

What I would expect to happen is that any attempt to invoke a member marked with an unrecognized/unsupported `modreq` should fail at runtime.  However, at the very least the binder could recognize and prevent this specific member access for `modreq(System.Runtime.CompilerServices.IsExternalInit)` so that C# "immutable" types with `init` properties cannot be mutated.

To reproduce:

1. Create a C# project with a class that has an `init` property, such as a record:

```cs
namespace RecordLib {
    public record Person {
        public string Name { get; init; }
    }
}
```

1. Create a VB.NET project referencing that C# project:

```vb
Option Strict Off

Imports RecordLib
Imports System

Module Program
    Sub Main()
        Dim person As New Person

        Dim obj As Object = person
        obj.Name = "Uh Oh!" ' member accessed via late binding

        Console.WriteLine(obj.Name) ' prints Uh Oh!
    End Sub
End Module
```

See also: https://github.com/dotnet/runtime/issues/40118

**Comment by VBAndCs** [comment]
I hope you don't fix this untill https://github.com/dotnet/roslyn/issues/49469 has a fix, as the late binding is the workaround I suggested to solve the issue there.
Don't complicate things for VB.NET that it is.
Thanks.

**Comment by danmoseley** [comment]
@cston we're branching for 7.0 on the 17th. Do you expect to fix this 6.0 bug by then?
